### DISCUSSION TITLE
Help tying gtk application / window into Qtile, as the desktop environment itself please

### DISCUSSION THREAD

**Discussion by Cameronsplaze** [qtile/qtile#4337] [Q&A]
I have a custom gtk.ApplicationWindowthat contains a tool bar, background, et all, that I'm trying to turn into a Desktop Environment. (Eventually I'll comment out qtiles bar. This gtk window will be the **only** thing on screen). I'm trying to find a way to make this gtk window be **unique** for every monitor/workspace (group), as in they're copies of the same object instead of all pointing to the same window instance, so they can each be in a different state.

This is what I have now, assume the parts I leave out are just the default config:

```python
from libqtile import bar, layout, widget, hook, qtile
from libqtile.config import Click, Drag, Group, Key, Match, Screen
from libqtile.lazy import lazy
from libqtile.utils import guess_terminal

import threading

gtk_app = main_application() # Gtk application to hook windows to

@hook.subscribe.startup
def autostart():
    # Thread it, so autostart is allowed to exit, but gtk_app is still running (it's a blocking call):
    thread = threading.Thread(target=gtk_app.run, args=(None,))
    thread.start()

screens = [
    Screen(
        bottom=bar.Bar(
            # Default settings for bar here....  ,
            window=main_window(parent=gtk_app, is_fullscreen=False), # Gtk applicationWindow instance (basically)
        ),
    ),
]
```

It won't start the window. I'm guessing it's because when starting the gtk app, there's no gtk windows tied into it *right then*, so it instantly quits for having no work.

BUT, if I comment out `window=` in Screen, then edit the gtk application itself to start one window, **it works** for the first group. That isn't scalable though, and doesn't even work for 2 groups.

Am I on the right track, of trying to get the window initialize into `Screen(window=*)`, and then somehow have each group create a unique screen instance for themselves? Or is there a cleaner way to do this? Thanks for any advice!

**Comment by elParaguayo** [comment]
I'm confused as to what you expect to happen here. I don't think `Screen` has a `window=` parameter (not does `Bar` which is where your code example suggests you're using it).

**Reply by Cameronsplaze** [reply]
Oh you're right, I can't find that param anywhere now either. Is a new `Screen` created for each `Group`? If so maybe I can subclass Screen, and add a key similar to `top` or `bottom`, but it takes up the whole screen instead?

**Reply by elParaguayo** [reply]
You can think of `Screen` as the physical monitor.

I think what you're after is a static window (which doesn't move with groups).

@m-col may have better suggestions about how you could achieve what you're trying to do but my concerns are that qtile will try to manage more than you want it to.

**Reply by Cameronsplaze** [reply]
Sounds good! I'd like a unique one per group, but tbh a static window would work amazingly if that's just not possible.

I just don't want to go down to the Wayland/X11 level myself for this project.

**Reply by Cameronsplaze** [reply]
I finally had time to read through some of the docs and source code a bit. The graph [here](https://qtile.readthedocs.io/en/stable/manual/commands/index.html#introduction-object-graph) helped a lot especially!

I'm thinking the solution might be to extend Groups? I can add onto the `Group.__init__` to create the window above and pin it to the group. (Hopefully there's some key in window that pins it to the background, so windows can't appear behind it). This way as the number of groups changes, each one still should have a "background and tool bar", so to speak.

But is there a reason [this group class](https://github.com/qtile/qtile/blob/master/libqtile/group.py#L42) begins with a underscore? I normally do that to tell other developers NOT to call/extend the object or method, since I might be removing it, or completely changing it's contents soon. Is it stable? I also see this [group class here](https://github.com/qtile/qtile/blob/master/libqtile/config.py#L674), but it has none of the window management methods you expect to see in a group, so I'm a little confused why there's two.
